### PR TITLE
Fix PrepareForUse potentially getting called twice on the same pooled usage

### PR DESCRIPTION
**Pull request by peppy**
This turns out to be quite dangerous and the cause for https://github.com/ppy/osu/issues/10938. It also fixes hitobjects disappearing in the editor at random.

Closes https://github.com/ppy/osu/issues/10938.

### CODE REVIEW

File: osu.Framework.Tests/Visual/Drawables/TestSceneDrawablePool.cs [hunk 1]
```diff
@@ -152,6 +152,30 @@ public void TestPrepareAndFreeMethods()
...
+            TestDrawable drawable = null;
+            TestDrawable drawable2 = null;
+
+            AddStep("consume item", () => drawable = consumeDrawable(false));
+
+            AddAssert("prepare was not run", () => drawable.PreparedCount == 0);
+            AddUntilStep("free was not run", () => drawable.FreedCount == 0);
+
+            AddStep("manually return drawable", () => pool.Return(drawable));
+            AddUntilStep("free was run", () => drawable.FreedCount == 1);
```
Comment (bdach) [inline osu.Framework.Tests/Visual/Drawables/TestSceneDrawablePool.cs:169]: This is about my only hang-up with this - should free be allowed to run if prepare hasn't? I could live with that being the case but I imagine that consumers might be operating on the premise that free can't run on drawables that were never prepared.

I suppose you could say that the state of the poolable before prepare is undefined in which case running free is probably better than not...

Comment (peppy) [reply osu.Framework.Tests/Visual/Drawables/TestSceneDrawablePool.cs:169]: It's actually *required* based on how we use these methods in some locations (take `FollowPointConnection` as an example):

```csharp
Get(() => { d.Entry = entry; } // run immediately in Assign

...

override FreeAfterUse()
{
    Entry = null;
}
```

Comment (smoogipoo) [reply osu.Framework.Tests/Visual/Drawables/TestSceneDrawablePool.cs:169]: How about leaving a remark on the xmldoc for `FreeAfterUse`?

This looks fine to me - I'm also suspicious, but I don't see a way of handling it without reconsidering this class - giving `PrepareForUse` an argument or having a `teardownAction` to complement `setupAction` in `Get()`. Both of those sound equally as bad and have their own downsides that actually probably cause more harm than good.

Comment (peppy) [reply osu.Framework.Tests/Visual/Drawables/TestSceneDrawablePool.cs:169]: Have added xmldoc for now.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**